runtime/near-vm-runner/src/cache.rs

```diff
@@ -28,41 +53,72 @@ pub(crate) fn compile_module_cached(
-        })?;
+    match cache.get(key.as_slice()) {
+        Ok(serialized) => {
+            match serialized {
+                Some(serialized) => {
+                    // We got cached code or error from DB cache.
+                    let record = CacheRecord::try_from_slice(serialized.as_slice()).unwrap();
+                    let code = match record {
+                        CacheRecord::Error(err) => return Err(err),
+                        CacheRecord::Code(code) => code,
+                    };
+                    let artifact = Artifact::deserialize(code.as_slice())
+                        .map_err(|_e| VMError::CacheError(DeserializationError))?;
+                    unsafe {
+                        load_cache_with(artifact, compiler_for_backend(Backend::Singlepass).unwrap().as_ref())
+                            .map_err(|_e| VMError::CacheError(DeserializationError))
+                    }
+                }
+                None => {
+                    // Nothing found in cache, create new record.
+                    let compiled: wasmer_runtime::Module =
+                        compile_module(code, config).map_err(|e| {
+                            let record = CacheRecord::Error(e.clone());
+                            let e1 = cache.put(key.as_slice(), &record.try_to_vec().unwrap());
+                            if e1.is_err() {
+                                // That's fine, just cannot cache compilation error.
+                                println!("Cannot cache an error");
+                            }
+                            e
+                        })?;
+
+                    let artifact = compiled.cache().map_err(|_e| {
+                        let e = VMError::CacheError(SerializationError);
+                        let record = CacheRecord::Error(e.clone());
+                        let e1 = cache.put(key.as_slice(), &record.try_to_vec().unwrap());
+                        if e1.is_err() {
+                            // That's fine, just cannot cache compilation error.
+                            println!("Cannot cache an error");
+                        }
+                        e
+                    })?;
 
-        let serialized = artifact.serialize().map_err(|_e| {
-            VMError::CacheError(format!("Cannot serialize cached artifact for {:?}", &code_hash))
-        })?;
-        cache.put(&hash, &serialized).map_err(|_e| {
-            VMError::CacheError(format!("Error saving cached artifact for {:?}", &code_hash))
-        })?;
-        Ok(compiled)
+                    let code = artifact.serialize().map_err(|_e| {
+                        let e = VMError::CacheError(SerializationError);
+                        let record = CacheRecord::Error(e.clone());
+                        cache.put(key.as_slice(), &record.try_to_vec().unwrap()).unwrap();
+                        e
+                    })?;
```

Comment:
Better to refactor this part into a separate method that has a return value of `Result<Code, CacheError>`.
This will allow you to write simple:
`let code = artifact.serialize().map_err(|_e| { VMError::CacheError(SerializationError) })?;`

Then at the end you can use:
```rust
let cache_record = match compile_and_serialize(...) {
  Ok(code) => CacheRecord::Code(),
  Err(e) => CacheRecord::Error(e),
}
```